File: C++/92-Reverse_Linked_List_II.cpp
// Runtime: 4 ms, faster than 72.39% of C++ online submissions for Reverse Linked List II.
// Memory Usage: 8.6 MB, less than 61.15% of C++ online submissions for Reverse Linked List II.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (!head)
            return NULL;
        ListNode* cur = head;
        ListNode* pre(NULL);
        while (m > 1) {
            pre = cur;
            cur = cur->next;
            --m;
            --n;
        }
        ListNode* tail = cur;
        ListNode* con = pre;
        while (n) {
            ListNode* tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
            --n;
        }
        if (con)
            con->next = pre;
        else
            head = pre;
        tail->next = cur;
        return head;
    }
};

File: Algorithms/Python3.x/92-Reverse_Linked_List_II.py
# Runtime: 36 ms, faster than 79.87% of Python3 online submissions for Reverse Linked List II.
# Memory Usage: 14 MB, less than 5.02% of Python3 online submissions for Reverse Linked List II.
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if not head:
            return None
        cur, pre = head, None
        while m > 1:
            pre = cur
            cur = cur.next
            m, n = m - 1, n - 1
        tail, con = cur, pre  # 2 pointer for the final connection
        while n:
            tmp = cur.next
            cur.next = pre
            pre = cur
            cur = tmp
            n -= 1
        if con:
            con.next = pre
        else:
            head = pre
        tail.next = cur
        return head
